Orange/widgets/utils/saveplot.py

```diff
@@ -0,0 +1,39 @@
+import os.path
+
+from PyQt4 import QtGui, QtCore
+
+
+# noinspection PyBroadException
+def save_plot(data, file_formats, filename=""):
+    writers = sorted(set(file_formats.values()), key=lambda w: w.PRIORITY)
+    filters = ['{} (*{})'.format(w.DESCRIPTION, ' *'.join(w.EXTENSIONS))
+               for w in writers]
+
+    _LAST_DIR_KEY = "directories/last_graph_directory"
+    _LAST_FMT_KEY = "directories/last_graph_format"
+    settings = QtCore.QSettings()
+    start_dir = settings.value(_LAST_DIR_KEY, filename)
+    if not start_dir or \
+            (not os.path.exists(start_dir) and
+             not os.path.exists(os.path.split(start_dir)[0])):
+        start_dir = os.path.expanduser("~")
+    last_ext = settings.value(_LAST_FMT_KEY, "")
+    if last_ext not in filters:
+        last_ext = filters[0]
+
+    filename, filter = QtGui.QFileDialog.getSaveFileNameAndFilter(
+        None, 'Save as ...', start_dir, ';;'.join(filters), last_ext)
+    if not filename:
+        return
+
+    writer = writers[filters.index(filter)]
+    if not os.path.splitext(filename)[1] and writer.EXTENSIONS:
+        filename += writer.EXTENSIONS[0]
+    try:
+        writer.write(filename, data)
```

Comment:
I agree that you may sometimes want to specify the format by typing an extension. But what if I'm intentionally specifying some particular non-standard extension? Say that the user sets the filter to svg, but then modifies the extension. If the extension is known, the filter is just (silently) overriden? This may be unexpected. And if the extension is unrecognized, the user could still have had a reason for this. When saving data, I may want to save in tab format, but as .txt because Excel recognizes it.

Both approaches make sense, so perhaps we should let the user make the decision (with another dialog) whether to (a) fix the extension, (b) use the format corresponding to the extension (your suggestion), (c) use the "wrong" combination (current solution) or (d) cancel?